### PR TITLE
Added reverse weibull fitting and allowing variations of translation value

### PR DESCRIPTION
**Pull request by akshay-raj-dhamija**

### CODE REVIEW

File: DistributionModels/weibull.py [hunk 1]
```diff
@@ -2,14 +2,18 @@
 import torch
 
 class weibull:
-    def __init__(self, saved_model=None):
+    def __init__(self, translateAmountTensor=1,saved_model=None):
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:5]: If `translateAmountTensor` is a float/int rather than a tensor this variable should be renamed.

File: DistributionModels/weibull.py [hunk 2]
```diff
@@ -2,14 +2,18 @@
...
+        self.smallScoreTensor = 0.0
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:6]: This variable should contain a tensor not float.

File: DistributionModels/weibull.py [hunk 3]
```diff
@@ -2,14 +2,18 @@
...
+        self.reversed = False
+        self.translateAmountTensor = translateAmountTensor
         if saved_model:
             self.wbFits = torch.zeros(saved_model['Scale'].shape[0],2)
             self.wbFits[:, 1] = saved_model['Scale']
             self.wbFits[:, 0] = saved_model['Shape']
             self.sign = saved_model['signTensor']
             self._ = saved_model['translateAmountTensor']
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:14]: This line should be removed since it was meant to ignore `translateAmountTensor`.

File: DistributionModels/weibull.py [hunk 4]
```diff
@@ -40,6 +44,25 @@ def FitLow(self,data, tailSize, isSorted=False, gpu=0):
         return to_return
 
 
+    def FitLowReversed(self,data, tailSize, isSorted=False, gpu=0):
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:47]: This function seems to simply set `self.reversed` flag, should this be absorbed in the existing `FitLow` function rather than a new function?

File: DistributionModels/weibull.py [hunk 5]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
         smallScoreTensor=self.smallScoreTensor
         if len(self.smallScoreTensor.shape)==2:
             smallScoreTensor=self.smallScoreTensor[:,0]
-        distances = distances + 1 - smallScoreTensor.to(self.deviceName)[None,:]
+            ## tb hack
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:110]: The hack is to change the `self.translateAmountTensor` tensor which is done at the constructor level and not here, so probably should remove this comment

File: DistributionModels/weibull.py [hunk 6]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
         smallScoreTensor=self.smallScoreTensor
         if len(self.smallScoreTensor.shape)==2:
             smallScoreTensor=self.smallScoreTensor[:,0]
-        distances = distances + 1 - smallScoreTensor.to(self.deviceName)[None,:]
+            ## tb hack 
+        distances = distances + self.translateAmountTensor - smallScoreTensor.to(self.deviceName)[None,:]
+#        distances = distances + 1 - smallScoreTensor.to(self.deviceName)[None,:]
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:112]: Remove the commented code here

File: DistributionModels/weibull.py [hunk 7]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
+        scale_tensor = self.wbFits[:,1]
+        shape_tensor = self.wbFits[:, 0]
+        if self.sign == -1:
+            distances = -distances
+        if len(distances.shape)==1:
+            distances = distances.repeat(shape_tensor.shape[0],1)
+        smallScoreTensor=self.smallScoreTensor
+        if len(self.smallScoreTensor.shape)==2:
+            smallScoreTensor=self.smallScoreTensor[:,0]
+            ## tb hack
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:141]: To remove

File: DistributionModels/weibull.py [hunk 8]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
+        if self.sign == -1:
+            distances = -distances
+        if len(distances.shape)==1:
+            distances = distances.repeat(shape_tensor.shape[0],1)
+        smallScoreTensor=self.smallScoreTensor
+        if len(self.smallScoreTensor.shape)==2:
+            smallScoreTensor=self.smallScoreTensor[:,0]
+            ## tb hack 
+        distances = distances + self.translateAmountTensor - smallScoreTensor.to(self.deviceName)[None,:]
+#        distances = distances + 1 - smallScoreTensor.to(self.deviceName)[None,:]
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:143]: To remove

File: DistributionModels/weibull.py [hunk 9]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
-                dataTensor = -dataTensor
-            sortedTensor = torch.topk(dataTensor, tailSize, dim=1, largest=True, sorted=True).values
+        if self.sign == -1:
+            dataTensor = -dataTensor
+
+#        if isSorted:
+#           could try to just pull top items but its complicated by
+#           sign and tesnor structor so for now just get topk 
+        sortedTensor = torch.topk(dataTensor, tailSize, dim=1, largest=True,
+                                      sorted=True).values
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:158]: Does this mean we do not have sorting capability now?

File: DistributionModels/weibull.py [hunk 10]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
 
         smallScoreTensor = sortedTensor[:, tailSize - 1].unsqueeze(1)
-        processedTensor = sortedTensor + 1 - smallScoreTensor
+        #tb hask to remove shift by small score, its not needed with proper use of translateAmountTensor
+#        smallScoreTensor = 0*smallScoreTensor;
+        processedTensor = sortedTensor + self.translateAmountTensor - smallScoreTensor
+# previously translateAmountTensor was hard coded at 1. (translateamount from libMR)
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:164]: To remove or be moved to constructor

File: DistributionModels/weibull.py [hunk 11]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
+        sortedTensor = torch.topk(dataTensor, tailSize, dim=1, largest=True,
+                                      sorted=True).values
 
         smallScoreTensor = sortedTensor[:, tailSize - 1].unsqueeze(1)
-        processedTensor = sortedTensor + 1 - smallScoreTensor
+        #tb hask to remove shift by small score, its not needed with proper use of translateAmountTensor
+#        smallScoreTensor = 0*smallScoreTensor;
+        processedTensor = sortedTensor + self.translateAmountTensor - smallScoreTensor
+# previously translateAmountTensor was hard coded at 1. (translateamount from libMR)        
+#        processedTensor = sortedTensor + 1 - smallScoreTensor
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:165]: To remove

File: DistributionModels/weibull.py [hunk 12]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
+#        if isSorted:
+#           could try to just pull top items but its complicated by
+#           sign and tesnor structor so for now just get topk 
+        sortedTensor = torch.topk(dataTensor, tailSize, dim=1, largest=True,
+                                      sorted=True).values
 
         smallScoreTensor = sortedTensor[:, tailSize - 1].unsqueeze(1)
-        processedTensor = sortedTensor + 1 - smallScoreTensor
+        #tb hask to remove shift by small score, its not needed with proper use of translateAmountTensor
+#        smallScoreTensor = 0*smallScoreTensor;
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:162]: Unsure if I understood these comments correctly, could you rephrase them?

File: DistributionModels/weibull.py [hunk 13]
```diff
@@ -84,27 +107,67 @@ def wscore(self, distances):
...
+        processedTensor = sortedTensor + self.translateAmountTensor - smallScoreTensor
+# previously translateAmountTensor was hard coded at 1. (translateamount from libMR)        
+#        processedTensor = sortedTensor + 1 - smallScoreTensor        
         # Returned in the format [Shape,Scale]
+        self.smallScoreTensor = smallScoreTensor
         if self.splits == 1:
             self.wbFits = self._fit(processedTensor)
-            self.smallScoreTensor = smallScoreTensor
+            return self.wbFits,smallScoreTensor
```
Comment (akshay-raj-dhamija) [inline DistributionModels/weibull.py:170]: Looks like a nice bug catch thanks @tboult, unsure if `self.smallScoreTensor` can be set outside in case `splits>1` @TouqeerAhmad could you please cross check

**Comment by akshay-raj-dhamija** [comment]
It seems like @TouqeerAhmad was once trying to have some tests at https://github.com/TouqeerAhmad/python_weibullfit_gpu in order to verify that the results match with libmr, does it make sense to have some test cases here so that finally the https://github.com/Vastlab/libMR repo can be archived since that repo gets a lot of setup issues.

File: DistributionModels/weibull.py (hunk 1 above)
Comment (siebenkopf) [reply DistributionModels/weibull.py:5]: Also, please stick to the pythonic way of naming variables, i.e., use `translate_amount_tensor` instead of the camelcase. Also, please **do not change** the order of parameters that are already there, otherwise a call to `weibull(model)` will not work any longer.

File: DistributionModels/weibull.py (hunk 4 above)
Comment (TouqeerAhmad) [reply DistributionModels/weibull.py:47]: I'm confused isn't reversed FitLow essentially FitHight
